content/blog/2025-10-20-zenoh-imoogi.md

```diff
@@ -0,0 +1,457 @@
+
+```rust
+#[repr(C)]
+struct MyData { x: u32, y: u16 }
+```
+
+you can build a layout and allocate a buffer for it, then wrap it in `Typed<MyData, …>` to access fields safely:
+
+```rust
+let layout = TypedLayout::<MyData>::new();
+let mut typed_buf: Typed<MyData, ZShmMut> = provider
+    .alloc(layout)
+    .wait()
+    .unwrap()
+    .assume_init();
+typed_buf.as_mut().x = 42;    // safe, typed access
+```
+
+Under the hood, this enforces the correct alignment and size for `MyData`, avoiding manual casting.
+
+### Improved Allocator
+
+The SHM allocator is now based on the [talc](https://crates.io/crates/talc) allocator and performs much more efficiently in real-world workload scenarios, addressing performance and fragmentation issues.
+
+### Improved Allocation Builder Ergonomics
+
+Allocation builders are now much simpler and more robust. The provider's `alloc` method is generic and accepts different layout descriptions such as `usize`, `(usize, AllocAlignment)`, and `MemoryLayout`.
+
+```rust
+// Option 1: Simple allocation
+let _shm_buf = provider.alloc(512).wait()?;
+// Option 2: Allocation with custom alignment
+let _shm_buf = provider
+            .alloc((512, AllocAlignment::ALIGN_2_BYTES))
+            .wait()?;
+/// Layout allocation:
+let layout = provider.alloc_layout(512)?;
+let _shm_buf = layout.alloc().wait()?;
+```
+
+
+
+### Buffer Layout, Resize and Safety
+
+The SHM API rework also clarified *owned vs. borrowed* buffers and added dynamic operations. Notably, you can now **resize a shared buffer in place**, which wasn't previously possible. For example, after allocating a buffer you can:
+
+```rust
+shm_buf.try_resize(new_len)?;
+```
+
+to grow or shrink it (within its segment's limits).
+
+We also expose alignment defaults and constants (`ALIGN_N_BYTE(S)`) and have tightened lifetime management: SHM buffers now own their segment handle to ensure proper cleanup. New unchecked mutability methods (`as_mut_unchecked()`) were added for cases where you *know* a mutable borrow is safe. Overall, these API changes make working with SHM buffers more predictable and idiomatic to Rust.
+
+### Implicit SHM Optimization
+
+Perhaps the most exciting change is that Zenoh now **automatically uses shared memory for large data**. In practice, this means you don't have to do anything special: if you publish a large payload, Zenoh will "implicitly pack" it into SHM and transmit just a reference locally. For example:
+
+```rust
+let large_data = vec![0u8; 1_000_000];   // 1 MB payload
+publisher.put(large_data).await.unwrap();
+// Zenoh will use SHM under the hood for this large message.
+```
+
+This implicit SHM transport applies in all modes (even when using a router), vastly improving throughput for local messages. In benchmark tests, the combined SHM changes provided approximately **10–100% throughput boost** (depending on message size).
```

Comment:
It would be worth to add a link to the configuration that controls the threshold on payload sizes for which SHM will be used as well as the configuration for the size of the SHM segment used by Zenoh.